platforms/platforms.go

```diff
@@ -0,0 +1,129 @@
+package platforms
+
+import (
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type platformKey struct {
+	os      string
+	arch    string
+	variant string
+}
+
+var (
+	selectorRe = regexp.MustCompile(`^[A-Za-z0-9_-]$`)
+)
+
+// ParseSelector parses the platform selector syntax into a platform
+// declaration.
+//
+// Platform selectors are in the format <os|arch>[/<arch>[/<variant>]]. The minimum
+// required information for a platform selector is the operating system.
+//
+// The minimum required information for a platform selector is the operating
+// system or architecture. If there is only a single string (no slashes(, the
+// value will be matched against the known set of operating systems, then fall
+// back to the known set of architectures. The missing component will be
+// inferred based on the local environment.
+func Parse(selector string) (specs.Platform, error) {
+	if strings.Contains(selector, "*") {
+		// TODO(stevvooe): need to work out exact wildcard handling
+		return specs.Platform{}, errors.Wrapf(errdefs.ErrInvalidArgument, "%q: wildcards not yet supported", selector)
+	}
+
+	if selectorRe.MatchString(selector) {
+		return specs.Platform{}, errors.Wrapf(errdefs.ErrInvalidArgument, "%q: platform selector must match %q", selector, selectorRe.String())
+	}
+
+	parts := strings.SplitN(selector, "/", 3)
+
+	var p specs.Platform
+	switch len(parts) {
+	case 1:
+		// in this case, we will test that the value might be an OS, then look
+		// it up. If it is not known, we'll treat it as an architecture. Since
+		// we have very little information about the platform here, we are
+		// going to be a little more strict if we don't know about the argument
+		// value.
+		p.OS = normalizeOS(parts[0])
+		if isKnownOS(p.OS) {
+			// picks a default architecture
+			p.Architecture = runtime.GOARCH
+			if p.Architecture == "arm" {
+				// TODO(stevvooe): Resolve arm variant, if not v6 (default)
+			}
+
+			return p, nil
+		}
+
+		p.Architecture, p.Variant = normalizeArch(parts[0], "")
+		if isKnownArch(p.Architecture) {
+			p.OS = runtime.GOOS
+			return p, nil
+		}
+
+		return specs.Platform{}, errors.Wrapf(errdefs.ErrInvalidArgument, "%q: unknown operating system or architecture", selector)
+	case 2:
+		// In this case, we treat as a regular os/arch pair. We don't care
+		// about whether or not we know of the platform.
+		p.OS = normalizeOS(parts[0])
+		p.Architecture, p.Variant = normalizeArch(parts[1], "")
+
+		return p, nil
+	case 3:
+		// we have a fully specified variant, this is rare
+		p.OS = normalizeOS(parts[0])
+		p.Architecture, p.Variant = normalizeArch(parts[1], parts[2])
+
+		return p, nil
+	}
+
+	return specs.Platform{}, errors.Wrapf(errdefs.ErrInvalidArgument, "%q: cannot parse platform selector", selector)
```

Comment:
I think this code is unreachable since `len(parts)` is always 1, 2, or 3, due to the way `strings.SplitN` works, not splitting after 3 parts. It might be better to use `strings.Split` since we should not allow `/` in a variant name.

Reply:
This is treated as the fall through case now.